### PR TITLE
Positivity-preserving limiter and SSPRK timestepping

### PR DESCRIPTION
**Pull request by thomasgibson**
**Questions for the review**:
- [ ] Is the scope and purpose of the PR clear?
  - [ ] The PR should have a description.
  - [ ] The PR should have a guide if needed (e.g., an ordering).
- [ ] Is every top-level method and class documented? Are things that should be documented actually so?
- [ ] Is the interface understandable? (I.e. can someone figure out what stuff does?) Is it well-defined?
- [ ] Does the implementation do what the docstring claims?
- [ ] Is everything that is implemented covered by tests?
- [ ] Do you see any immediate risks or performance disadvantages with the design? Example: what do interface normals attach to?

### CODE REVIEW

File: mirgecom/integrators/ssprk.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+def ssprk43_step(state, t, dt, rhs, limiter=None):
+    """Take one step using an explicit 4-stage, 3rd-order, SSPRK method."""
+    actx = state.array_context
+
+    def rhs_update(t, y):
+        return y + dt*rhs(t, y)
+
+    y1 = 1/2*state + 1/2*rhs_update(t, state)
+    if limiter is not None:
+        y1 = thaw(freeze(limiter(y1), actx), actx)
```
Comment (inducer) [inline mirgecom/integrators/ssprk.py:42]: Move the `thaw(freeze(...))` inside the limiter. Has nothing to do with time stepping.

File: mirgecom/integrators/ssprk.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+def ssprk43_step(state, t, dt, rhs, limiter=None):
+    """Take one step using an explicit 4-stage, 3rd-order, SSPRK method."""
+    actx = state.array_context
+
+    def rhs_update(t, y):
+        return y + dt*rhs(t, y)
+
+    y1 = 1/2*state + 1/2*rhs_update(t, state)
+    if limiter is not None:
```
Comment (inducer) [inline mirgecom/integrators/ssprk.py:41]: Ditch the `if limiter is None` for:
```py
if limiter is None:
    limiter = lambda x: x
```

File: mirgecom/limiter.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    actx = state.array_context
+
+    def compute_limited_field(field):
+        if not isinstance(field, DOFArray):
+            # vecs is not a DOFArray -> treat as array container
+            return map_array_container(compute_limited_field, field)
+
+        # Compute nodal and elementwise max/mins of the field
+        _mmax = op.nodal_max(dcoll, "banana", field)
+        _mmin = op.nodal_min(dcoll, "avocado", field)
```
Comment (inducer) [inline mirgecom/limiter.py:72]: :banana: :avocado:

File: mirgecom/limiter.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+from meshmode.dof_array import DOFArray
+
+from grudge.discretization import DiscretizationCollection
+
+from mirgecom.fluid import ConservedVars
+
+import grudge.op as op
+
+
+def limiter_liu_osher(dcoll: DiscretizationCollection, state):
```
Comment (inducer) [inline mirgecom/limiter.py:39]: Take in two `dd`s, and two `state`s:
- One on which the nodal minima/maxima are taken.
- One that actually has the limiting applied.

File: mirgecom/limiter.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            )
+        )
+        return theta*(field - cell_avgs) + cell_avgs
+
+    return ConservedVars(
+        mass=compute_limited_field(state.mass),
+        energy=state.energy,
+        momentum=state.momentum,
+        species_mass=compute_limited_field(state.species_mass)
+    )
```
Comment (inducer) [inline mirgecom/limiter.py:98]: Make it apply to single scalars and/or generic array containers.

File: mirgecom/limiter.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        An array container containing the filtered field(s).
+    """
+    from grudge.geometry import area_element
+
+    actx = state.array_context
+
+    def compute_limited_field(field):
+        if not isinstance(field, DOFArray):
+            # vecs is not a DOFArray -> treat as array container
+            return map_array_container(compute_limited_field, field)
```
Comment (inducer) [inline mirgecom/limiter.py:68]: use `multi_map_array_container`

File: mirgecom/limiter.py (hunk 4 above)
Comment (inducer) [inline mirgecom/limiter.py:39]: The base limiter might be a candidate for grudge?

**Comment by MTCam** [comment]
I like this limiter, but would like to suggest some changes along the lines of #600. This change set puts the limiting into the hands of the user so they can choose what fields to limit.   Consider that the original function is incorrectly limiting both mass and species_mass, and offers no user control over what is limited.   

The change set suggested in #600 also should work with any time-integrator.   Note the added combustion example in `autoignition-mpi.py`, and allowing driver control allows the two different limited fields in the combustion case and the `pulse-mpi.py` case.

**Pull request closed without merging**